cpp/include/cuspatial/experimental/detail/trajectory_distances_and_speeds.cuh

```diff
@@ -0,0 +1,137 @@
+
+  auto id_point_timestamp = thrust::make_zip_iterator(ids_first, points_first, timestamps_first);
+
+  rmm::device_uvector<Rep> durations(num_points, stream);
+  rmm::device_uvector<T> distances(num_points, stream);
+
+  // initialize just the first elements since we will skip them
+  durations.set_element_to_zero_async(0, stream);
+  distances.set_element_to_zero_async(0, stream);
+
+  auto duration_and_distance = thrust::make_zip_iterator(durations.begin(), distances.begin());
+
+  thrust::transform(rmm::exec_policy(stream),
+                    id_point_timestamp,
+                    id_point_timestamp + num_points - 1,
+                    id_point_timestamp + 1,
+                    duration_and_distance + 1,
+                    [] __device__(auto const& p0, auto const& p1) {
+                      if (thrust::get<0>(p0) == thrust::get<0>(p1)) {  // ids are the same
+                        Point pos0   = thrust::get<1>(p0);
+                        Point pos1   = thrust::get<1>(p1);
+                        Timestamp t0 = thrust::get<2>(p0);
+                        Timestamp t1 = thrust::get<2>(p1);
+                        Point vec    = pos1 - pos0;
+
+                        return thrust::make_tuple((t1 - t0).count(), sqrt(dot(vec, vec)));
+                      }
+                      return thrust::make_tuple(Rep{}, T{});
+                    });
+
+  auto duration_and_distance_tmp = thrust::make_zip_iterator(durations.begin(),
+                                                             distances.begin(),
+                                                             thrust::make_constant_iterator<T>(0),
+                                                             thrust::make_constant_iterator<T>(0));
+
+  rmm::device_uvector<Rep> durations_tmp(num_trajectories, stream);
+  rmm::device_uvector<T> distances_tmp(num_trajectories, stream);
+
+  auto distances_begin = thrust::get<0>(distances_and_speeds_first.get_iterator_tuple());
+  auto speeds_begin    = thrust::get<1>(distances_and_speeds_first.get_iterator_tuple());
+
+  auto duration_distances_and_speed = thrust::make_zip_iterator(
+    durations_tmp.begin(), distances_tmp.begin(), distances_begin, speeds_begin);
+
+  using Period =
+    typename cuda::std::ratio_divide<typename Timestamp::period, typename Sec::period>::type;
+
+  // Reduce the intermediate durations and kilometer distances into meter
+  // distances and speeds in meters/second
+  thrust::reduce_by_key(rmm::exec_policy(stream),
+                        ids_first,  // keys
+                        ids_last,
+                        duration_and_distance_tmp,       // values
+                        thrust::discard_iterator(),      // keys_output
+                        duration_distances_and_speed,    // values_output
+                        thrust::equal_to<Id>(),          // binary_pred
+                        [] __device__(auto a, auto b) {  // binary_op
```

Comment:
The trailing comments can be removed. Depending on the IDE one uses, the name of the function arguments might be shown as UI.